src/trio/_core/_exceptions.py

```diff
@@ -67,11 +73,49 @@ class Cancelled(BaseException, metaclass=NoPublicConstructor):
 
     """
 
+    source: Literal["deadline", "nursery", "explicit", "unknown", "KeyboardInterrupt"]
+    # TODO: this should probably be a Task?
+    source_task: str | None = None
+    reason: str | None = None
+
     def __str__(self) -> str:
-        return "Cancelled"
+        def repr_if_not_none(lead: str, s: str | None, trail: str = "") -> str:
+            if s is None:
+                return ""
+            return lead + s + trail
+
+        return (
+            f"cancelled due to {self.source}"
+            + repr_if_not_none(" with reason '", self.reason, "'")
+            + repr_if_not_none(" from task ", self.source_task)
+        )
 
     def __reduce__(self) -> tuple[Callable[[], Cancelled], tuple[()]]:
-        return (Cancelled._create, ())
+        # the `__reduce__` tuple does not support kwargs, so we must use partial
+        # for non-default args
+        # or switch to allow posarg (?)
```

Comment:
I'd rather have this once in `__reduce__` than worse signatures at every call site.

Reply:
`__reduce__` can return more than just the two elements -- the third item is either a tuple (in which case it gets passed to `__setstate__`) or a dict (in which it gets added to `__dict__`).

Either of those options should work.

Reply:
> `__reduce__` can return more than just the two elements -- the third item is either a tuple (in which case it gets passed to `__setstate__`) or a dict (in which it gets added to `__dict__`).
> 
> Either of those options should work.

neither does because he have a required kwarg. If it's tuple we're passing posarg, if it's a dict we're modifying attributes after initializing - but `source` is required.

Normally people would get around this by using [`__getnewargs_ex__`](https://docs.python.org/3/library/pickle.html#object.__getnewargs_ex__) but since we *also* need `__reduce__` to specify the callable object we either have to use both, or do partial+`__reduce__`.

(not fully confident this is 100% correct, this is all messy)

Reply:
We control both the callable passed, and how `__setstate__` works, and both are effectively private API. The tuple could just have a variable number of elements, and then we pass a private module-level function to do the unpickling?

Reply:
well, `partial` inside `__reduce__` works fine, so I don't think we need to bother with more complicated stuff.